src/lib/mdns/DiscoveryManager.cpp

```diff
@@ -330,5 +331,81 @@ void DiscoveryManager::HandleNodeIdResolve(void * context, MdnsService * result,
 #endif // CHIP_ENABLE_MDNS
 }
 
+#if CHIP_ENABLE_MDNS
+CHIP_ERROR DiscoveryManager::BrowseUnprovisionedDevice(Inet::IPAddressType type, Inet::InterfaceId interface)
+{
+    return ChipMdnsBrowse("_chipc", MdnsServiceProtocol::kMdnsProtocolUdp, type, interface, HandleUnprovisionedDeviceBrowse, this);
+}
+
+void DiscoveryManager::HandleUnprovisionedDeviceBrowse(void * context, MdnsService * services, size_t servicesSize,
+                                                       CHIP_ERROR error)
+{
+    DiscoveryManager * mgr = static_cast<DiscoveryManager *>(context);
+
+    if (mgr->mResolveDelegate)
+    {
+        mgr->mResolveDelegate->HandleUnprovisionedDeviceBrowse(error, services, servicesSize);
+    }
+}
+
+CHIP_ERROR DiscoveryManager::ResolveUnprovisionedDevice(uint64_t name, Inet::IPAddressType type, Inet::InterfaceId interface)
+{
+    MdnsService service;
+
+    snprintf(service.mName, sizeof(service.mName), "%016" PRIX64, name);
+    strncpy(service.mType, "_chipc", sizeof(service.mType));
+    service.mProtocol    = MdnsServiceProtocol::kMdnsProtocolUdp;
+    service.mAddressType = type;
+    return ChipMdnsResolve(&service, interface, HandleDiscriminatorResolve, this);
+}
+
+void DiscoveryManager::HandleDiscriminatorResolve(void * context, MdnsService * result, CHIP_ERROR error)
+{
+    DiscoveryManager * mgr = static_cast<DiscoveryManager *>(context);
+
+    if (mgr->mResolveDelegate == nullptr)
+    {
+        return;
+    }
+    if (error != CHIP_NO_ERROR)
+    {
+        ChipLogError(Discovery, "Node ID resolved failed with %s", chip::ErrorStr(error));
+        mgr->mResolveDelegate->HandleUnprovisionedDeviceResolve(error, 0, MdnsService{});
+    }
+    else if (result == nullptr)
+    {
+        ChipLogError(Discovery, "Node ID resolve not found");
+        mgr->mResolveDelegate->HandleUnprovisionedDeviceResolve(CHIP_ERROR_UNKNOWN_RESOURCE_ID, 0, MdnsService{});
+    }
+    else
+    {
+        uint64_t name = 0;
+
+        for (size_t i = 0; i < sizeof(uint64_t) * 2; i++)
```

Comment:
I think I've seen similar code elsewhere, can we write a utility for this?  I think it should also verify that it's not indexing off the end of result...